### PR TITLE
Add new conditions pipeline stages

### PR DESCRIPTION
**Pull request by matthchr**
Add pipeline stage to put conditions on all status types.

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

**Comment by matthchr** [comment]
Ah, looks like this has to be rebase onto #1688

I will leave this as is for now and rebase it once 1688 makes it to master. Comments still welcome on this even though CI is failing (it should pass after rebase)

File: hack/generator/pkg/astmodel/conditioner_interface.go
Comment (matthchr) [inline hack/generator/pkg/astmodel/conditioner_interface.go:20]: Methods missing comments in here...

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (matthchr) [inline hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:228]: maybe clarify this if statement a bit

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:222]:
```suggestion
// conditionsPropertyHandler generates conversions for the "Conditions" status property. This property is set by the controller
```

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:223]: "currently" implies that maybe it should in future? Is this right?
```suggestion
// after each reconcile and does not need to be preserved.
```

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:232]: Returning an empty statement feels like a bit of a hack, esp given it's not idiomatic Go to return sentinel values. Consider a later refactoring where each handler returns `([]dst.Stmt, bool)`?

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/codegen/pipeline/add_status_conditions.go:21]:
```suggestion
		"Adds the property 'Conditions' to all status types and implements genruntime.Conditioner on all resources",
```

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/codegen/pipeline/add_status_conditions.go:51]: I think you can do this more simply using the `Overlay()` method on `Types`.

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/codegen/pipeline/add_status_conditions.go:64]: Use `PropertyInjector` to do the heavy lifting

File: hack/generator/pkg/codegen/pipeline/add_status_conditions_test.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/codegen/pipeline/add_status_conditions_test.go:38]:
```suggestion
	// and that the conditions.Conditioner interface is properly implemented on the resource.
```

**Review by theunrepentantgeek — COMMENTED** [review]
You've said it needs work (including rebase) so I won't approve, but looks good.

gif

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:222]: Good eye

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:223]: There's a theoretical future where we have support for more than just the `Ready` condition, and other conditions aren't set in each reconcile loop pass possibly?

It seems an unlikely future though, especially because if we don't rederive the conditions each pass there are cases (such as when `Status` gets lost) where we might be making assumptions about what we have and not updating them - have made the change you suggested as it more clearly explains the _now_, and we can leave the future for the future.

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:232]: I've made a note to do this but will do it in a later PR. Given that the infrastructure that uses this function signature is shared across to/from ARM (and there are a number of methods in each) it'll be a pretty big refactor.

Agree it'd make it a bit clearer though.

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (matthchr) [reply hack/generator/pkg/codegen/pipeline/add_status_conditions.go:21]: Fixed

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (matthchr) [reply hack/generator/pkg/codegen/pipeline/add_status_conditions.go:51]: Yep, you're right. Moved to use that.

File: hack/generator/pkg/codegen/pipeline/add_status_conditions_test.go
Comment (matthchr) [reply hack/generator/pkg/codegen/pipeline/add_status_conditions_test.go:38]: Fixed

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (matthchr) [reply hack/generator/pkg/codegen/pipeline/add_status_conditions.go:64]: That is nice!

The only problem is that I want to make sure to error if we're trying to add a property that already exists. It seems like maybe `PropertyInjector` should have that behavior by default too? Otherwise we'll be silently overwriting properties.

I'll delete my code here and update `PropertyInjector` to do that.

**Comment by matthchr** [comment]
I've rebased this now so it should be good to merge once approved

File: hack/generator/pkg/astmodel/property_injector.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/property_injector.go:32]: Maybe include the name of the property too (in case the error has another cause)?
```suggestion
		return TypeDefinition{}, errors.Wrapf(err, "failed to inject property %q into %q", prop.Name(), def.Name())
```

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/codegen/pipeline/add_status_conditions.go:19]: Ideally new stages wouldn't use `MakeLegacyStage()`, they'd use `MakeStage()` instead.

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/codegen/pipeline/add_status_conditions.go:46]:
```suggestion
					return nil, errors.Wrapf(err, "couldn't create genruntime.Conditioner implementation for %q", def.Name())
```

**Review by theunrepentantgeek — APPROVED** [review]
Looks good.

gif

File: hack/generator/pkg/astmodel/property_injector.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/property_injector.go:32]: Fixed

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (matthchr) [reply hack/generator/pkg/codegen/pipeline/add_status_conditions.go:19]: Fixed

File: hack/generator/pkg/codegen/pipeline/add_status_conditions.go
Comment (matthchr) [reply hack/generator/pkg/codegen/pipeline/add_status_conditions.go:46]: Fixed

**Pull request merged**
